core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -194,35 +212,37 @@ class TransactionCoordinator(brokerId: Int,
                                        pid: Long,
                                        epoch: Short,
                                        partitions: collection.Set[TopicPartition],
-                                       responseCallback: TxnMetadataUpdateCallback): Unit = {
+                                       responseCallback: AddPartitionsCallback): Unit = {
     val errors = validateTransactionalId(transactionalId)
     if (errors != Errors.NONE)
       responseCallback(errors)
     else {
       // try to update the transaction metadata and append the updated metadata to txn log;
       // if there is no such metadata treat it as invalid pid mapping error.
-      val (error, newMetadata) = txnManager.getTransactionState(transactionalId) match {
+      val (error: Errors, coordinatorEpoch: Int, txnMetadata: TransactionMetadata) = txnManager.getTransactionState(transactionalId) match {
```

Comment:
Similar to a comment that I left in a different PR, adding type annotations here triggers a pattern match and it's brittle if there are cases where the match can fail, but the compiler can't prove it (any time there is a subclass involved). It's one of the cases where relying on type inference is safer (it will always infer a safe type).